### PR TITLE
[FEATURE] Clear caches on invalidation for embedded resources

### PR DESCRIPTION
**Pull request by e0ipso**
If a resource was cached in a denormalized response as an embedded
resource, and that resource needs to be invalidated; we also need to
clear all the resources containing it.

### CODE REVIEW

**Comment by amitaibu** [comment]
Needs review?

**Comment by e0ipso** [comment]
Not just yet

**Comment by e0ipso** [comment]
This is ready to be reviewed @amitaibu.

**Comment by amitaibu** [comment]
Cool, I just need to take a deep breath before ;)

File: restful.entity.inc [hunk 1]
```diff
@@ -96,9 +86,11 @@ function restful_entity_update($entity, $type) {
...
       return;
     }
-    $handler
-      ->getCacheController()
-      ->clear($hash);
+    // You can get away without the fragments for a clear.
+    $cache_object = new RenderCache(new ArrayCollection(), $hash, $handler->getCacheController());
+    // Do a clear with the RenderCache object to also remove the cache fragment
+    // entities.
+    $cache_object->clear();
```
Comment (e0ipso) [inline restful.entity.inc:93]: I was not clearing the render cache properly before. If you do it this way, you clear the caches and delete the involved cache fragments.

File: src/Formatter/FormatterManager.php [hunk 2]
```diff
@@ -212,9 +212,9 @@ public function getPlugin($instance_id) {
   protected function getPluginByName($name) {
     /* @var FormatterInterface $formatter */
     $formatter = $this->plugins->get($name);
-    $formatter->setConfiguration(array(
-      'resource' => $this->resource,
-    ));
+    if ($this->resource) {
+      $formatter->setResource($this->resource);
+    }
```
Comment (e0ipso) [inline src/Formatter/FormatterManager.php:217]: This was a problem for tests. The resource plugin was stored and could not be changed for the subsequent test. This was not an issue during regular processing.

File: src/Plugin/formatter/Formatter.php [hunk 3]
```diff
@@ -162,26 +207,84 @@ protected function createCacheController($data) {
...
+   *
+   * @param bool|string[] $allowed_fields
+   *   The list of allowed fields in dot notation.
+   * @param string $prefix
+   *   The prefix used to select the fields and to remove from the front.
+   *
+   * @return bool|string[]
+   *   The new allowed fields for the nested sub-request.
+   */
+  protected static function unprefixInputOptions($allowed_fields, $prefix) {
```
Comment (e0ipso) [inline src/Plugin/formatter/Formatter.php:275]: Given some query param like `a,b,b.c,b.d,e,f,f.g,f.h,f.i`, this is used so we can pass only the request options `b.c,b.d` when making a sub-request to the resource for the `b` relationship.

**Comment by amitaibu** [comment]
I will review on Thursday - marked in my calendar :)

**Comment by e0ipso** [comment]
That's great!

Let me know if you want to go through it together.

**Pull request merged**